It looks like for me a minimal working reproducible example of the runasp test goes something like:
1) Make a test environment on one of our linux machines and pip install the test ska_shell into it (PYTHONPATH doesn't work for the old code still using Ska.Shell)
2) Source into the new environment (needs "production" environment to run arc5gl for runasp)
3) Make a clean clone of runasp and run a test obsid (this is the one that failed for me though I assume all fail the same).
```
cd /tmp
git clone git@github.com:sot/runasp
cd runasp
mkdir obs30739
cd obs30739
python ../runasp.py --obsid 30739
```
And that fails with 
```
2025-02-19 14:54:05,815 punlearn: Command not found.
2025-02-19 14:54:05,816 Tcsh-2025-02-19T14:54:05.81>
Traceback (most recent call last):
  File "/tmp/runasp/obs30739/../runasp.py", line 695, in <module>
    main(opt)
  File "/tmp/runasp/obs30739/../runasp.py", line 680, in main
    run_ai(range_ais)
  File "/tmp/runasp/obs30739/../runasp.py", line 396, in run_ai
    loglines = tcsh_shell("punlearn asp_l1_std",
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/fido.real/miniforge3/envs/test/lib/python3.12/site-packages/Ska/Shell/shell.py", line 238, in tcsh_shell
    outlines, newenv = run_shell(
                       ^^^^^^^^^^
  File "/fido.real/miniforge3/envs/test/lib/python3.12/site-packages/Ska/Shell/shell.py", line 153, in run_shell
    raise exc
Ska.Shell.shell.ShellError: Shell error: punlearn: Command not found.
```

--- a/ska_shell/shell.py
+++ b/ska_shell/shell.py
@@ -15,6 +15,10 @@ class ShellError(Exception):
     pass
 
 
+class NonZeroReturnCode(ShellError):
+    pass
+
+
 def _fix_paths(envs, pathvars=('PATH', 'PERLLIB', 'PERL5LIB', 'PYTHONPATH',
                                'LD_LIBRARY_PATH', 'MANPATH', 'INFOPATH')):
     """For the specified env vars that represent a search path, make sure that the
@@ -148,7 +152,7 @@ def run_shell(cmdstr, shell='bash', logfile=None, importenv=False, getenv=False,
         logfile.write(f"{shell.capitalize()}-{time}>\n")
     if proc.returncode:
         msg = " ".join(stdout[-1:])  # stdout could be empty
-        exc = ShellError(f"Shell error: {msg}")
+        exc = NonZeroReturnCode(f"Shell error: {msg}")
         exc.lines = stdout
         raise exc
 

--- a/ska_shell/shell.py
+++ b/ska_shell/shell.py
@@ -135,6 +135,10 @@ def run_shell(cmdstr, shell='bash', logfile=None, importenv=False, getenv=False,
     # all lines are joined so the shell exits at the first failure
     cmdstr = " && ".join([c for c in cmdstr.splitlines() if c.strip()])
 
+    if shell in ["tcsh", "csh"]:
+        shell = "bash"
+        cmdstr = f"{shell} -f -c '{cmdstr}'"
+
     proc = subprocess.Popen(
         [cmdstr],
         executable=shell,

--- a/ska_shell/shell.py
+++ b/ska_shell/shell.py
@@ -136,8 +136,8 @@ def run_shell(cmdstr, shell='bash', logfile=None, importenv=False, getenv=False,
     cmdstr = " && ".join([c for c in cmdstr.splitlines() if c.strip()])
 
     if shell in ["tcsh", "csh"]:
-        shell = "bash"
         cmdstr = f"{shell} -f -c '{cmdstr}'"
+        shell = "bash"
 
     proc = subprocess.Popen(
         [cmdstr],

